### PR TITLE
feat(primitives, storage): save prune checkpoints in database

### LINKED ISSUE #3432: Save pruning checkpoints in a separate database table

### PR DESCRIPTION
**Pull request by shekhirin**
Resolves https://github.com/paradigmxyz/reth/issues/3432

Not a breaking change: we've added new table. Also, we're free to change it without breaking changes until we start writing into it.

### CODE REVIEW

File: crates/storage/db/src/tables/models/mod.rs [hunk 1]
```diff
@@ -135,3 +136,20 @@ impl Decode for StoredNibblesSubKey {
...
 }
+
+impl Encode for PrunePart {
+    type Encoded = [u8; 1];
+
+    fn encode(self) -> Self::Encoded {
+        let mut bytes = BytesMut::with_capacity(1);
+        self.to_compact(&mut bytes);
+        bytes.as_ref().try_into().expect("incorrect length")
+    }
```
Comment (shekhirin) [inline crates/storage/db/src/tables/models/mod.rs:147]: @joshieDo there surely is a better way to do that but I can't figure that out.

The most obvious way
```rust
let mut buf = [0u8];
self.to_compact(&mut buf);
buf
```
doesn't work:
```console
error[E0277]: the trait bound `[u8; 1]: BufMut` is not satisfied
   --> crates/storage/db/src/tables/models/mod.rs:145:25
    |
145 |         self.to_compact(&mut buf);
    |              ---------- ^^^^^^^^ the trait `BufMut` is not implemented for `[u8; 1]`
    |              |
    |              required by a bound introduced by this call
    |
    = help: the trait `BufMut` is implemented for `&mut [u8]`
note: required by a bound in `to_compact`
   --> /Users/shekhirin/Projects/reth/crates/storage/codecs/src/lib.rs:32:12
    |
32  |         B: bytes::BufMut + AsMut<[u8]>;
    |            ^^^^^^^^^^^^^ required by this bound in `Compact::to_compact`
help: convert the array to a `&mut [u8]` slice instead
    |
145 |         self.to_compact(&mut buf[..]);
    |                                 ++++
```

Following the compiler suggestion and doing 
```rust
let mut buf = [0u8];
self.to_compact(&mut buf[..]);
buf
```
also doesn't work:
```console
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
   --> crates/storage/db/src/tables/models/mod.rs:145:25
    |
145 |         self.to_compact(&mut buf[..]);
    |              ---------- ^^^^^^^^^^^^ doesn't have a size known at compile-time
    |              |
    |              required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `[u8]`
note: required by a bound in `to_compact`
   --> /Users/shekhirin/Projects/reth/crates/storage/codecs/src/lib.rs:30:19
    |
30  |     fn to_compact<B>(self, buf: &mut B) -> usize
    |                   ^ required by this bound in `Compact::to_compact`
```

---

I suspect there's something in how we restrict the `B` in `to_compact` but don't understand straight away.

Comment (shekhirin) [reply crates/storage/db/src/tables/models/mod.rs]: solved

File: crates/primitives/src/prune/mode.rs [hunk 2]
```diff
@@ -17,7 +17,7 @@ pub enum PruneMode {
 #[cfg(test)]
```
Comment (mattsse) [inline crates/primitives/src/prune/mode.rs:17]: ah we only need this for tests?

otherwise can use #[default]

**Review by mattsse — APPROVED** [review]
this only defines another table and types and doesn't write to/create it.

perhaps we also want proptest for this?

lgtm otherwise, pending @joshieDo

File: crates/primitives/src/prune/mode.rs (hunk 2 above)
Comment (shekhirin) [reply crates/primitives/src/prune/mode.rs:17]: yea, only for that part: https://github.com/paradigmxyz/reth/blob/73eeca0e29dda19cc9b481020c5b669fdaaff5e1/crates/storage/codecs/derive/src/compact/generator.rs#L31-L34

**Comment by shekhirin** [comment]
@mattsse `main_codec` generates prop tests for both

<details>
    <summary>PrunePart</summary>

```rust
#[allow(non_snake_case)]
#[cfg(test)]
mod PrunePartTests {
    use super::Compact;
    extern crate test;
    #[cfg(test)]
    #[rustc_test_marker = "prune::part::PrunePartTests::proptest"]
    pub const proptest: test::TestDescAndFn = test::TestDescAndFn {
        desc: test::TestDesc {
            name: test::StaticTestName("prune::part::PrunePartTests::proptest"),
            ignore: false,
            ignore_message: ::core::option::Option::None,
            source_file: "crates/primitives/src/prune/part.rs",
            start_line: 4usize,
            start_col: 1usize,
            end_line: 4usize,
            end_col: 14usize,
            compile_fail: false,
            no_run: false,
            should_panic: test::ShouldPanic::No,
            test_type: test::TestType::UnitTest,
        },
        testfn: test::StaticTestFn(|| test::assert_test_result(proptest())),
    };
    fn proptest() {
        let mut config = proptest::prelude::ProptestConfig::with_cases(256i32 as u32);
        {
            let mut config = config.__sugar_to_owned();
            ::proptest::sugar::force_no_fork(&mut config);
            {
                config.source_file = Some("crates/primitives/src/prune/part.rs");
                let mut runner = ::proptest::test_runner::TestRunner::new(config);
                let names = "field";
                match runner.run(
                    &::proptest::strategy::Strategy::prop_map(
                        ::proptest::arbitrary::any::<super::PrunePart>(),
                        |values| ::proptest::sugar::NamedArguments(names, values),
                    ),
                    |::proptest::sugar::NamedArguments(_, field)| {
                        let _: () = {
                            {
                                let mut buf = ::alloc::vec::Vec::new();
                                let len = field.clone().to_compact(&mut buf);
                                let (decoded, _) = super::PrunePart::from_compact(&buf, len);
                                if !(field == decoded) {
                                    ::core::panicking::panic(
                                        "assertion failed: field == decoded",
                                    )
                                };
                            }
                        };
                        Ok(())
                    },
                ) {
                    Ok(_) => (),
                    Err(e) => {
                        ::std::rt::panic_fmt(format_args!("{0}\n{1}", e, runner));
                    }
                }
            };
        };
    }
}
```
</details>

<details>
    <summary>PruneCheckpoint</summary>

```rust
#[allow(non_snake_case)]
#[cfg(test)]
mod PruneCheckpointTests {
    use super::Compact;
    extern crate test;
    #[cfg(test)]
    #[rustc_test_marker = "prune::checkpoint::PruneCheckpointTests::proptest"]
    pub const proptest: test::TestDescAndFn = test::TestDescAndFn {
        desc: test::TestDesc {
            name: test::StaticTestName("prune::checkpoint::PruneCheckpointTests::proptest"),
            ignore: false,
            ignore_message: ::core::option::Option::None,
            source_file: "crates/primitives/src/prune/checkpoint.rs",
            start_line: 5usize,
            start_col: 1usize,
            end_line: 5usize,
            end_col: 14usize,
            compile_fail: false,
            no_run: false,
            should_panic: test::ShouldPanic::No,
            test_type: test::TestType::UnitTest,
        },
        testfn: test::StaticTestFn(|| test::assert_test_result(proptest())),
    };
    fn proptest() {
        let mut config = proptest::prelude::ProptestConfig::with_cases(256i32 as u32);
        {
            let mut config = config.__sugar_to_owned();
            ::proptest::sugar::force_no_fork(&mut config);
            {
                config.source_file = Some("crates/primitives/src/prune/checkpoint.rs");
                let mut runner = ::proptest::test_runner::TestRunner::new(config);
                let names = "field";
                match runner.run(
                    &::proptest::strategy::Strategy::prop_map(
                        ::proptest::arbitrary::any::<super::PruneCheckpoint>(),
                        |values| ::proptest::sugar::NamedArguments(names, values),
                    ),
                    |::proptest::sugar::NamedArguments(_, field)| {
                        let _: () = {
                            {
                                let mut buf = ::alloc::vec::Vec::new();
                                let len = field.clone().to_compact(&mut buf);
                                let (decoded, _) =
                                    super::PruneCheckpoint::from_compact(&buf, len);
                                if !(field == decoded) {
                                    ::core::panicking::panic(
                                        "assertion failed: field == decoded",
                                    )
                                };
                            }
                        };
                        Ok(())
                    },
                ) {
                    Ok(_) => (),
                    Err(e) => {
                        ::std::rt::panic_fmt(format_args!("{0}\n{1}", e, runner));
                    }
                }
            };
        };
    }
}
```
</details>

**Review by joshieDo — APPROVED** [review]

**Pull request merged**